cli/azd/pkg/project/service_target_aks.go

```diff
@@ -0,0 +1,444 @@
+		},
+		func(service *kubectl.Service) bool {
+			// If the service is not a load balancer it should be immediately available
+			if service.Spec.Type != kubectl.ServiceTypeLoadBalancer {
+				return true
+			}
+
+			// Load balancer can take some time to be provision by AKS
+			var ipAddress string
+			for _, config := range service.Status.LoadBalancer.Ingress {
+				if config.Ip != "" {
+					ipAddress = config.Ip
+					break
+				}
+			}
+
+			return ipAddress != ""
+		},
+	)
+}
+
+// Retrieve any service endpoints for the specified namespace and serviceNameFilter
+// Supports service types for LoadBalancer and ClusterIP
+func (t *aksTarget) getServiceEndpoints(ctx context.Context, namespace string, serviceNameFilter string) ([]string, error) {
+	service, err := t.waitForService(ctx, namespace, serviceNameFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	var endpoints []string
+	if service.Spec.Type == kubectl.ServiceTypeLoadBalancer {
+		for _, resource := range service.Status.LoadBalancer.Ingress {
+			endpoints = append(endpoints, fmt.Sprintf("http://%s (Service, Type: LoadBalancer)", resource.Ip))
+		}
+	} else if service.Spec.Type == kubectl.ServiceTypeClusterIp {
+		for index, ip := range service.Spec.ClusterIps {
+			endpoints = append(endpoints, fmt.Sprintf("http://%s:%d (Service, Type: ClusterIP)", ip, service.Spec.Ports[index].Port))
+		}
+	}
+
+	return endpoints, nil
+}
+
+// Retrieve any ingress endpoints for the specified namespace and serviceNameFilter
+// Supports service types for LoadBalancer, supports Hosts and/or IP address
+func (t *aksTarget) getIngressEndpoints(ctx context.Context, namespace string, resourceFilter string) ([]string, error) {
+	ingress, err := t.waitForIngress(ctx, namespace, resourceFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	var endpoints []string
+	var protocol string
+	if ingress.Spec.Tls == nil {
+		protocol = "http"
+	} else {
+		protocol = "https"
+	}
+
+	for index, resource := range ingress.Status.LoadBalancer.Ingress {
+		var baseUrl string
+		if ingress.Spec.Rules[index].Host == nil {
+			baseUrl = fmt.Sprintf("%s://%s", protocol, resource.Ip)
+		} else {
+			baseUrl = fmt.Sprintf("%s://%s", *ingress.Spec.Rules[index].Host, resource.Ip)
+		}
+
+		endpointUrl, err := url.JoinPath(baseUrl, t.config.K8s.Ingress.RelativePath)
+		if err != nil {
+			return nil, fmt.Errorf("failed constructing service endpoints, %w", err)
+		}
+
+		endpoints = append(endpoints, fmt.Sprintf("%s  (Ingress, Type: LoadBalancer)", endpointUrl))
+	}
+
+	return endpoints, nil
+}
+
+func (t *aksTarget) generateImageTag() (string, error) {
+	configuredTag, err := t.config.Docker.Tag.Envsubst(t.env.Getenv)
+	if err != nil {
+		return "", err
+	}
+
+	if configuredTag != "" {
+		return configuredTag, nil
+	}
+
+	return fmt.Sprintf("%s/%s-%s:azdev-deploy-%d",
+		strings.ToLower(t.config.Project.Name),
+		strings.ToLower(t.config.Name),
+		strings.ToLower(t.env.GetEnvName()),
+		t.clock.Now().Unix(),
```

Comment:
nit. Consider doing IoC here and injecting an implementation for tagging?  That way we can mock the tag in tests with a deterministic tag.  Right now, the tag is different every second in time.